### PR TITLE
Threaded General User Objects

### PR DESCRIPTION
**Pull request by andrsd**
Allow creating threaded copies of general user objects

### CODE REVIEW

**Comment by andrsd** [comment]
@jasondhales Can you test this branch with your BISON changes to see if that fixes the threading problem you described in #11731? Thanks!

**Comment by jasondhales** [comment]
If I change "threaded_copies" to "threaded" in FluidProperties.C, it works.

**Comment by andrsd** [comment]
Thanks for catching that!

File: framework/src/loops/ComputeThreadedGeneralUserObjectsThread.C [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+void
+ComputeThreadedGeneralUserObjectsThread::operator()(const EmptyRange & /*range*/,
+                                                    bool bypass_threading)
+{
+  try
+  {
+    ParallelUniqueId puid;
+    _tid = bypass_threading ? 0 : puid.id;
+
+    const auto & objects = _user_objects.getActiveObjects();
```
Comment (friedmud) [inline framework/src/loops/ComputeThreadedGeneralUserObjectsThread.C]: I don't get it - won't this make every thread execute every object?

File: framework/src/problems/FEProblemBase.C [hunk 2]
```diff
@@ -2955,6 +2971,18 @@ FEProblemBase::computeUserObjects(const ExecFlagType & type, const Moose::AuxGro
   // Finalize, threadJoin, and update PP values of Nodal
   finalizeUserObjects<NodalUserObject>(nodal);
 
+  // Initialize threaded GeneralUserObjects
+  initializeUserObjects<GeneralUserObject>(threaded_general);
+
+  if (threaded_general.hasActiveObjects())
+  {
+    ComputeThreadedGeneralUserObjectsThread ctguot(*this, threaded_general);
+    Threads::parallel_reduce(EmptyRange(), ctguot);
```
Comment (friedmud) [inline framework/src/problems/FEProblemBase.C]: This is what I was saying we shouldn't do.  You need to actually pass the range of userobjects here to ensure they all get executed.

**Review by friedmud — CHANGES_REQUESTED** [review]
Some issues with execution

**Comment by rwcarlsen** [comment]
This makes me sad - I will have to rewrite this code and resolve lots of conflicts :-/

**Comment by jasondhales** [comment]
I'm not sure why this is an issue for @rwcarlsen, but I'm eager to see this merged.

**Review by friedmud — CHANGES_REQUESTED** [review]
The code here looks good - but I would like to see one test of it.  Can you dream up something that would test this?

**Comment by andrsd** [comment]
> The code here looks good - but I would like to see one test of it. Can you dream up something that would test this?

I flipped the "threaded" flag for the `FluidProperty` class, so this would be tested whenever we run tests in the `fluid_property` module. If that is not enough, I could build an user object that has some state, so that when it runs threaded and does not have the `threaded` flag on, it would produce a wrong result. If we get "lucky". Other than that, I am not sure how else to test this.

**Comment by rwcarlsen** [comment]
It would be helpful if we have a reliable test (that fails without this
fix/feature) - if we don't I'm likely to introduce regressions soon when I
rework the code.

**Comment by andrsd** [comment]
The problem is that the result would depend on the thread scheduler which we have no control over, so the test could sometimes pass even if there was something wrong :-(

**Comment by friedmud** [comment]
The thread scheduler doesn't matter here - every threaded object will get
executed... even if not on separate threads.  All we need to verify is that
each one gets executed...

And, yes, I would like a reliable test in moose_test before this can go
in.  It's a tricky enough capability that we need to make sure it stays
working correctly.

**Comment by andrsd** [comment]
(Hopefully) reliable test added...

**Comment by permcody** [comment]
ping @friedmud? 8 days no action

File: framework/src/problems/FEProblemBase.C [hunk 3]
```diff
@@ -2955,6 +2972,22 @@ FEProblemBase::computeUserObjects(const ExecFlagType & type, const Moose::AuxGro
...
+  initializeUserObjects<GeneralUserObject>(threaded_general);
+
+  if (threaded_general.hasActiveObjects())
+  {
+    std::vector<THREAD_ID> thread_ids(libMesh::n_threads());
+    for (THREAD_ID tid = 0; tid < libMesh::n_threads(); ++tid)
+      thread_ids[tid] = tid;
+
+    ComputeThreadedGeneralUserObjectsThread ctguot(*this, threaded_general);
+    Threads::parallel_reduce(ThreadIdRange(thread_ids.begin(), thread_ids.end()), ctguot);
```
Comment (friedmud) [inline framework/src/problems/FEProblemBase.C]: Ok - the new comment is that this doesn't work with interdependencies.  Each ThreadedGeneralUserObject needs its own thread loop - then call `threadjoin()` and `finalize()` on it before starting the next one.

**Review by friedmud — CHANGES_REQUESTED** [review]
Issue with interdependencies between these objects.

**Review by friedmud — APPROVED** [review]
👍

**Comment by brianmoose** [comment]
Looks like our minimum clang testing has a problem with this PR: https://civet.inl.gov/job/218243/

**Comment by andrsd** [comment]
#11961 should fix that

**Pull request merged**